protocol/block.go

```diff
@@ -63,58 +69,86 @@ func (c *Chain) GenerateBlock(ctx context.Context, prev *bc.Block, snapshot *sta
 		},
 	}
 
+	var txEntries []*bc.TxEntries
+
 	for _, tx := range txs {
 		if len(b.Transactions) >= maxBlockTxs {
 			break
 		}
 
-		// TODO(jackson): Should this go in ConfirmTx too?
-		err = c.checkIssuanceWindow(tx)
+		// Filter out transactions that are not well-formed.
+		err := c.ValidateTx(tx.TxEntries)
 		if err != nil {
+			// TODO(bobg): log this?
 			continue
 		}
 
-		if validation.ConfirmTx(result, c.InitialBlockHash, bc.NewBlockVersion, timestampMS, tx) == nil {
-			err = validation.ApplyTx(result, tx)
-			if err != nil {
-				return nil, nil, err
-			}
-			b.Transactions = append(b.Transactions, tx)
+		// Filter out double-spends etc.
+		err = newSnapshot.ApplyTx(tx.TxEntries)
+		if err != nil {
+			// TODO(bobg): log this?
+			continue
 		}
+
+		b.Transactions = append(b.Transactions, tx)
+		txEntries = append(txEntries, tx.TxEntries)
 	}
-	b.TransactionsMerkleRoot, err = bc.MerkleRoot(b.Transactions)
+
+	var err error
+
+	b.TransactionsMerkleRoot, err = bc.MerkleRoot(txEntries)
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "calculating tx merkle root")
 	}
-	b.AssetsMerkleRoot = result.Tree.RootHash()
-	return b, result, nil
+
+	b.AssetsMerkleRoot = newSnapshot.Tree.RootHash()
+
+	return b, newSnapshot, nil
+}
+
+// ValidateBlock validates an incoming block in advance of applying it
+// to a snapshot (with ApplyValidBlock) and committing it to the
+// blockchain (with CommitAppliedBlock).
+func (c *Chain) ValidateBlock(block, prev *bc.Block) error {
+	return c.validateBlock(block, prev, true)
+}
+
+func (c *Chain) validateBlock(block, prev *bc.Block, runProg bool) error {
```

Comment:
Yeah something like that. I think it can be made a little cleaner by fiddling with the runProg parameter as @jbowens and I have suggested. Honestly I'm happy with this PR either way. It will be easier for us to discuss and evaluate these alternatives separately.